### PR TITLE
cmd/build: Add buildah support

### PR DESCRIPTION
**Pull request by mkimuram**
**Description of the change:**
This PR adds buildah support.

**Motivation for the change:**
operator-sdk should support building image by using buildah. 
There is an existing discussion for motivation in https://github.com/operator-framework/operator-sdk/pull/563 .

### CODE REVIEW

**Comment by mkimuram** [comment]
It seems that https://github.com/operator-framework/operator-sdk/pull/563 is doing a similar change, but it has no update for a long time.
I guess that the reason that the PR takes much time to merge would be that the PR includes big changes on other features like multi-stage builds and adding e2e tests to use buildah everywhere, therefore I hope that this PR helps making progress on just supporting buildah.
(Actually, I had a trouble in building operator in my only-buildah-installed environment, so adding this feature early will help me a lot.)

For memcached-operator, builds with below commands succeeded after applying this PR,
in an envrionment where buildah is installed.

```
$ export IMAGE=quay.io/example-inc/memcached-operator:v0.0.1
$ operator-sdk build --image-builder buildah $IMAGE
$ sed "s@REPLACE_IMAGE@$IMAGE@g" -i deploy/operator.yaml
$ operator-sdk build --image-builder buildah --enable-tests $IMAGE
```

@estroz @AlexNPavel @shawn-hurley 

Could you review this PR?

File: cmd/operator-sdk/build/cmd.go [hunk 1]
```diff
@@ -63,6 +65,8 @@ For example:
 	buildCmd.Flags().StringVar(&testLocationBuild, "test-location", "./test/e2e", "Location of tests")
 	buildCmd.Flags().StringVar(&namespacedManBuild, "namespaced-manifest", "deploy/operator.yaml", "Path of namespaced resources manifest for tests")
 	buildCmd.Flags().StringVar(&dockerBuildArgs, "docker-build-args", "", "Extra docker build arguments as one string such as \"--build-arg https_proxy=$https_proxy\"")
+	buildCmd.Flags().StringVar(&buildahBuildArgs, "buildah-build-args", "", "Extra buildah build arguments as one string such as \"--build-arg https_proxy=$https_proxy\"")
```
Comment (shawn-hurley) [inline cmd/operator-sdk/build/cmd.go]: WDYT of re-using a flag `build-args` for both docker and buildah?

Comment (mkimuram) [reply cmd/operator-sdk/build/cmd.go]: Can we rename ```docker-build-args``` with something like ```image-build-args```?

Then re-using flag makes sense to me. I'm not sure how operator-sdk community thinks about the backward compatibility of command line option.

Comment (estroz) [reply cmd/operator-sdk/build/cmd.go]: `image-build-args` is fine with me. We'll have to add these changes to the `CHANGELOG.md` too.

File: cmd/operator-sdk/build/cmd.go [hunk 2]
```diff
@@ -172,15 +176,31 @@ func buildFunc(cmd *cobra.Command, args []string) error {
...
+			dbArgs = append(dbArgs, splitArgs...)
+		}
+
+		buildCmd = exec.Command("docker", dbArgs...)
+	case "buildah":
+		bbArgs := []string{"bud", "--format=docker", "-f", "build/Dockerfile", "-t", baseImageName, "."}
+
+		if buildahBuildArgs != "" {
+			splitArgs := strings.Fields(buildahBuildArgs)
+			bbArgs = append(bbArgs, splitArgs...)
```
Comment (estroz) [inline cmd/operator-sdk/build/cmd.go]: You'll need to append `buildahBuildArgs` before `.`, as `buildah` doesn't allow args to come after build context.

File: cmd/operator-sdk/build/cmd.go [hunk 3]
```diff
@@ -234,15 +254,31 @@ func buildFunc(cmd *cobra.Command, args []string) error {
...
+				testDbArgs = append(testDbArgs, splitArgs...)
+			}
+
+			testBuildCmd = exec.Command("docker", testDbArgs...)
+		case "buildah":
+			testBbArgs := []string{"bud", "--format=docker", "-f", testDockerfile, "-t", image, "--build-arg", "NAMESPACEDMAN=" + namespacedManBuild, "--build-arg", "BASEIMAGE=" + baseImageName, "."}
+
+			if buildahBuildArgs != "" {
+				splitArgs := strings.Fields(buildahBuildArgs)
+				testBbArgs = append(testBbArgs, splitArgs...)
```
Comment (estroz) [inline cmd/operator-sdk/build/cmd.go]: Same as above.

File: cmd/operator-sdk/build/cmd.go [hunk 4]
```diff
@@ -172,15 +176,31 @@ func buildFunc(cmd *cobra.Command, args []string) error {
 
 	log.Infof("Building Docker image %s", baseImageName)
 
-	dbArgs := []string{"build", ".", "-f", "build/Dockerfile", "-t", baseImageName}
+	var buildCmd *exec.Cmd
+	switch imageBuilder {
+	case "docker":
```
Comment (estroz) [inline cmd/operator-sdk/build/cmd.go]: Refactor this and the test image build switch statements into functions.

**Comment by mkimuram** [comment]
@shawn-hurley @estroz 

Thank you for your feedback. I've fixed as suggested. PTAL

File: cmd/operator-sdk/build/cmd.go [hunk 5]
```diff
@@ -141,6 +143,27 @@ func verifyTestManifest(image string) error {
 	return nil
 }
 
+func createBuildCommand(imageBuilder, context, dockerFile, image, imageBuildArgs string) (*exec.Cmd, error) {
```
Comment (estroz) [inline cmd/operator-sdk/build/cmd.go]: `imageBuildArgs ...string` so we can pass any number of string args without having to create a new string from them. Then you can loop over `imageBuildArgs` like so:
```go
for _, bargs := range imageBuildArgs {
	if bargs != "" {
		splitArgs := strings.Fields(bargs)
		args = append(args, splitArgs...)
	}
}
```

File: cmd/operator-sdk/build/cmd.go [hunk 6]
```diff
@@ -232,17 +252,15 @@ func buildFunc(cmd *cobra.Command, args []string) error {
...
 
-		log.Infof("Building test Docker image %s", image)
+		log.Infof("Building test OCI image %s", image)
 
-		testDbArgs := []string{"build", ".", "-f", testDockerfile, "-t", image, "--build-arg", "NAMESPACEDMAN=" + namespacedManBuild, "--build-arg", "BASEIMAGE=" + baseImageName}
-
-		if dockerBuildArgs != "" {
-			splitArgs := strings.Fields(dockerBuildArgs)
-			testDbArgs = append(testDbArgs, splitArgs...)
+		testImageBuildArgs := fmt.Sprintf("%s --build-arg NAMESPACEDMAN=%s --build-arg BASEIMAGE=%s", imageBuildArgs, namespacedManBuild, baseImageName)
```
Comment (estroz) [inline cmd/operator-sdk/build/cmd.go]:
```suggestion
		testImageBuildArgs := fmt.Sprintf("--build-arg NAMESPACEDMAN=%s --build-arg BASEIMAGE=%s", namespacedManBuild, baseImageName)
```
After changing `createBuildCommand`'s signature.

File: cmd/operator-sdk/build/cmd.go [hunk 7]
```diff
@@ -232,17 +252,15 @@ func buildFunc(cmd *cobra.Command, args []string) error {
...
+		testBuildCmd, err := createBuildCommand(imageBuilder, ".", testDockerfile, image, testImageBuildArgs)
```
Comment (estroz) [inline cmd/operator-sdk/build/cmd.go]:
```suggestion
		testBuildCmd, err := createBuildCommand(imageBuilder, ".", testDockerfile, image, imageBuildArgs, testImageBuildArgs)
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,8 +2,12 @@
 
 ### Added
 
+- New option for [`operator-sdk build --image-builder`](https://github.com/operator-framework/operator-sdk/blob/master/doc/sdk-cli-reference.md#build), which can be used to specify which image builder to use. ([#1311](https://github.com/operator-framework/operator-sdk/pull/1311))
```
Comment (estroz) [inline CHANGELOG.md]:
```suggestion
- New option for [`operator-sdk build --image-builder`](https://github.com/operator-framework/operator-sdk/blob/master/doc/sdk-cli-reference.md#build), which can be used to specify which image builder to use. Adds support for [buildah](https://github.com/containers/buildah/). ([#1311](https://github.com/operator-framework/operator-sdk/pull/1311))
```

**Comment by mkimuram** [comment]
@estroz 

Thank you for your suggestion. Fixed as suggested again. PTAL

**Comment by mkimuram** [comment]
Below error happens without quote for `[docker, buildah]`, so fixed in the last commit.

```
./hack/ci/marker --root=doc
Found broken reference (docker, buildah -> docker, buildah) at doc/sdk-cli-reference.md:20
make: *** [test/markdown] Error 1
```

**Comment by mkimuram** [comment]
Rebased again to resolve conflict. PTAL

**Pull request merged**